### PR TITLE
Ensure that when an async read is performed on a response content the errors are correctly dealt with.

### PR DESCRIPTION
**Pull request by mandel-macaque**
Some network errors in iOS are exposed as Stream errors, specially those
related with the lost of the connection. This patch ensures that if such
an issue happens an exception is raised in the async read task.

Fixes bug https://bugzilla.xamarin.com/show_bug.cgi?id=53792

### CODE REVIEW

File: mcs/class/System.Net.Http/CFContentStream.cs [hunk 1]
```diff
@@ -54,6 +57,7 @@ class CFContentStream : HttpContent
 		Mutex data_mutex;
 		AutoResetEvent data_event;
 		AutoResetEvent data_read_event;
+		Exception http_exception = null;
```
Comment (marek-safar) [inline mcs/class/System.Net.Http/CFContentStream.cs]: Redundant null initializaiton

File: mcs/class/System.Net.Http/CFContentStream.cs [hunk 2]
```diff
@@ -85,6 +90,17 @@ public CFContentStream (CFHTTPStream stream)
 			data_mutex = new Mutex ();
 		}
 
+		void HandleErrorEvent (object sender, CFStream.StreamEventArgs e)
+		{
+			var gotMutex = data_mutex.WaitOne ();
+			if (gotMutex) {
+				var stream = (CFHTTPStream)sender;
+				if (e.EventType == CFStreamEventType.ErrorOccurred)
+					http_exception = stream.GetError ();
```
Comment (marek-safar) [inline mcs/class/System.Net.Http/CFContentStream.cs]: Better to use ExceptionDispatchInfo.Capture

Comment (marek-safar) [inline mcs/class/System.Net.Http/CFContentStream.cs]: Use Volatile.Write

File: mcs/class/System.Net.Http/CFContentStream.cs (hunk 1 above)
Comment (marek-safar) [inline mcs/class/System.Net.Http/CFContentStream.cs]: Redundant null initialization

File: mcs/class/System.Net.Http/CFContentStream.cs [hunk 3]
```diff
@@ -77,6 +81,7 @@ class CFContentStream : HttpContent
 		public CFContentStream (CFHTTPStream stream)
 		{
 			this.http_stream = stream;
+			this.http_stream.ErrorEvent += HandleErrorEvent;
```
Comment (marek-safar) [inline mcs/class/System.Net.Http/CFContentStream.cs]: Close method should do -=

File: mcs/class/System.Net.Http/CFContentStream.cs [hunk 4]
```diff
@@ -111,6 +127,10 @@ public void Close ()
 		{
 			while (data_event.WaitOne ()) {
 				data_mutex.WaitOne ();
+				if (http_exception != null) {
+					data_mutex.ReleaseMutex ();
+					throw http_exception;
```
Comment (marek-safar) [inline mcs/class/System.Net.Http/CFContentStream.cs]: Use ExceptionDispatchInfo.Throw intead

**Comment by marek-safar** [comment]
It'd be nice to add a test for this to HttpClientTest.cs

File: mcs/class/System.Net.Http/CFContentStream.cs (hunk 3 above)
Comment (mandel-macaque) [inline mcs/class/System.Net.Http/CFContentStream.cs]: Done in the last commit. Of course github does no close the comment 'cause is not that smart.

**Pull request merged**